pkg/image/registryclient/client.go

```diff
@@ -223,7 +248,48 @@ func (c *Context) Repository(ctx context.Context, registry *url.URL, repoName st
 	if limiter == nil {
 		limiter = rate.NewLimiter(rate.Limit(5), 5)
 	}
-	return NewLimitedRetryRepository(repo, c.Retries, limiter), nil
+	retryRepo := NewLimitedRetryRepository(repo, c.Retries, limiter)
+	if c.AlternateRepositoryStrategy != nil {
+		retryRepo.alternateStrategy = c.AlternateRepositoryStrategy
+	}
+	return retryRepo, nil
+}
+
+// RepositoryWithAlternateReference returns the next cached repository and ref, or will gather alternate sources then return first repo, ref, or error
+// Once the repository is returned it is removed from the cache.
+func (c *Context) RepositoryWithAlternateReference(ctx context.Context, locator imagereference.DockerImageReference, cachedRepos map[imagereference.DockerImageReference]distribution.Repository) (distribution.Repository, imagereference.DockerImageReference, error) {
+	repo, err := c.repository(ctx, locator.DockerClientDefaults().RegistryURL(), locator.RepositoryName(), true)
+	if repo.alternateStrategy == nil {
```

Comment:
I am porting `image-registry` to use this patch and I am seeing the following panic here:

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
```

That is because `repo` is `nil`.

Reply:
I added a check for if retryRepository.Repository == nil, return an error (if no alternative image strategy). I think that is where the nil panic was coming from? you were returned a nil Repository, then tried to use it? If not, I'll take a closer look, get more information from you. ie, was the retryRepository nil or the returned retryRepository.Repository?

Reply:
got it, thanks

Reply:
Thanks @sallyom. That addressed a problem with our tests in `image-registry`

Reply:
great, thank you for the update

Reply:
and I updated that just now to only return if repo==nil, since that's what is needed (otherwise will skip looking at alt sources if err != nil && repo != nil